### PR TITLE
Add endpoint for creating/editing/deleting floating IPs

### PR DESCRIPTION
**Pull request by lfu**
```
POST /api/floating_ips/10000000000054
{
    "action": "edit",
    "status": "test"
}
```
RESPONSE
```
{
    "success": true,
    "message": "Updating Floating Ip id:10000000000054 name: '10.9.60.154'",
    "task_id": "10000000053300",
    "task_href": "http://localhost:3000/api/tasks/10000000053300"
}
```

https://github.com/ManageIQ/manageiq-api/issues/901

@miq-bot add_label enhancement

### CODE REVIEW

File: app/controllers/api/floating_ips_controller.rb [hunk 1]
```diff
@@ -1,4 +1,39 @@
 module Api
   class FloatingIpsController < BaseController
+    def create_resource(_type, _id = nil, data = {})
+      ems = ExtManagementSystem.find(data['ems_id'])
+      raise BadRequestError, "Create floating ip not supported for Provider: #{ems.name}" unless ems.respond_to?(:create_floating_ip_queue)
```
Comment (agrare) [inline app/controllers/api/floating_ips_controller.rb:5]: We should use `SupportsFeatureMixin` for this, e.g. `ems.supports?(:create_floating_ip)

File: app/controllers/api/floating_ips_controller.rb [hunk 2]
```diff
@@ -1,4 +1,39 @@
...
+
+      task_id = ems.create_floating_ip_queue(session[:userid], data.deep_symbolize_keys)
+      action_result(true, "Creating Floating Ip #{data['name']} for Provider: #{ems.name}", :task_id => task_id)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def edit_resource(type, id, data)
+      floating_ip = resource_search(id, type, collection_class(:floating_ips))
+      raise BadRequestError, "Update not supported for #{floating_ip_ident(floating_ip)}" unless floating_ip.respond_to?(:update_floating_ip_queue)
```
Comment (agrare) [inline app/controllers/api/floating_ips_controller.rb:15]: `FloatingIp` already has `floating_ip.supports?(:update)`

File: app/controllers/api/floating_ips_controller.rb [hunk 3]
```diff
@@ -1,4 +1,39 @@
...
+      task_id = floating_ip.update_floating_ip_queue(session[:userid], data.deep_symbolize_keys)
+      action_result(true, "Updating #{floating_ip_ident(floating_ip)}", :task_id => task_id)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def delete_resource(type, id, _data = {})
+      delete_action_handler do
+        floating_ip = resource_search(id, type, collection_class(:floating_ips))
+        raise BadRequestError, "Delete not supported for #{floating_ip_ident(floating_ip)}" unless floating_ip.respond_to?(:delete_floating_ip_queue)
```
Comment (agrare) [inline app/controllers/api/floating_ips_controller.rb:26]: Same, `floating_ip.supports?(:delete)` should work

File: app/controllers/api/floating_ips_controller.rb (hunk 1 above)
Comment (agrare) [reply app/controllers/api/floating_ips_controller.rb:5]: You can also raise `.unsupported_reason(:feature)` for each of these

File: app/controllers/api/floating_ips_controller.rb [hunk 4]
```diff
@@ -1,4 +1,39 @@
...
+      ems = ExtManagementSystem.find(data['ems_id'])
+      raise BadRequestError, "Create floating ip not supported for Provider: #{ems.name}" unless ems.respond_to?(:create_floating_ip_queue)
+
+      task_id = ems.create_floating_ip_queue(session[:userid], data.deep_symbolize_keys)
+      action_result(true, "Creating Floating Ip #{data['name']} for Provider: #{ems.name}", :task_id => task_id)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def edit_resource(type, id, data)
```
Comment (agrare) [inline app/controllers/api/floating_ips_controller.rb:13]: Re: rubocop warnings...I feel like `id` should be an allowed variable name

File: app/controllers/api/floating_ips_controller.rb [hunk 5]
```diff
@@ -1,4 +1,40 @@
 module Api
   class FloatingIpsController < BaseController
+    def create_resource(_type, _id = nil, data = {})
```
Comment (Fryguy) [inline app/controllers/api/floating_ips_controller.rb:3]: Does this come from a PUT or from a POST with "create" action (or both)?  What does the incoming payload look like and does it match the shape of the GET request?

**Review by Fryguy — APPROVED** [review]
LGTM - @agrare Can you review the changes (specifically the supports feature stuff?)

File: app/controllers/api/floating_ips_controller.rb [hunk 6]
```diff
@@ -1,4 +1,40 @@
...
+
+      task_id = ems.create_floating_ip_queue(session[:userid], data.deep_symbolize_keys)
+      action_result(true, "Creating Floating Ip #{data['name']} for Provider: #{ems.name}", :task_id => task_id)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def edit_resource(type, id, data)
+      floating_ip = resource_search(id, type, collection_class(:floating_ips))
+      raise BadRequestError, "Update for #{floating_ip_ident(floating_ip)}: #{floating_ip.unsupported_reason(:update)}" unless floating_ip.supports?(:update)
```
Comment (agrare) [inline app/controllers/api/floating_ips_controller.rb:16]: I just noticed that the supports feature is `:update` but the unsupported reason is `:update_floating_ip` [[ref]](https://github.com/ManageIQ/manageiq-providers-openstack/blob/master/app/models/manageiq/providers/openstack/network_manager/floating_ip.rb#L18) so you likely won't see the proper error here but that looks like a bug on the openstack side not here

**Review by agrare — APPROVED** [review]
:+1: LGTM

File: spec/requests/floating_ips_spec.rb [hunk 7]
```diff
@@ -53,5 +53,148 @@
...
+    it "forbids multiple floating ip deletion without an appropriate role" do
+      floating_ip1, floating_ip2 = FactoryBot.create_list(:floating_ip, 2)
+      api_basic_authorize
+
+      post(api_floating_ips_url, :params => {:action => "delete", :resources => [{:id => floating_ip1.id}, {:id => floating_ip2.id}]})
+
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    it 'raises an error when delete not supported for floating ip' do
```
Comment (abellotti) [inline spec/requests/floating_ips_spec.rb:186]: minor, quoting not consistent with other tests, ' vs "

**Review by gtanzillo — APPROVED** [review]

**Pull request merged**